[6.5] UI port airgun discovered hosts tests

ldjebran opened the pull request:
```bash
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/ui_airgun/test_discoveredhost.py 
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: file
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.2.1, mock-1.6.3, forked-0.2, cov-2.5.1
collecting 10 items                                                                                          2018-09-06 13:21:47 - conftest - DEBUG - BZ deselect is disabled in settings

collected 10 items                                                                                           

tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_read_details_from_discovered_host PASSED [ 10%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_pxe_based_discovery PASSED               [ 20%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_pxe_less_with_dhcp_unattended PASSED     [ 30%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_provision_using_quick_host_button PASSED [ 40%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_name PASSED                       [ 50%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_manual_provision_host_with_rule PASSED   [ 60%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_delete PASSED                            [ 70%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_default_org PASSED                [ 80%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_default_location PASSED           [ 90%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_reboot PASSED                            [100%]

======================================== 10 passed in 2122.60 seconds ========================================
```
dependent from airgun PR: 
put DO NOT MERGE waiting for airgun PR to be merged https://github.com/SatelliteQE/airgun/pull/202

rplevka commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,496 @@
+@run_only_on('sat')
+@skip_if_not_set('compute_resources', 'vlan_networking')
+@tier3
+def test_positive_update_name(
```
this is what my original comment was about - let;s simply create a `make_fake_discovery_host` factory method and use it here (we use similar `make_fake_host` already in other host-related tests).
This will instantly make the test a T2 tests which will be much faster and will test the same thing - note that we are not interacting with the real host here in any way. We simply create the host jsut so it shows up in the discovered hosts - but we can mock this easily.
- if you want, i can write the factory for you.

ldjebran replied:
@rplevka How we can provision a fake host?
Note: we are provisioning by updating the discovered host, provisioning data by changing only the name.

Please clarify: """we use similar make_fake_host already in other host-related tests""", in which test we are using fake host for provisioning.

ldjebran replied:
@rplevka I understand that we can create a test discovered host but we are in tier3 and we need a real user scenario. Do not see why we have to cut some tests as soon as we have already all the setups already done in this module, the run time is adequate here, but the value is very high.

rplevka replied:
@ldjebran because we are not interested in HOW the discovered host got into satellite here. the test says that the host should be already discovered and it only tries the renaming part.
this should not be tier3 at all (see my first comment on this, which suggests to make this tier2).

- we have the proper discovery covered in other tests.
- making the full-blown discovery just to test we can do some CRUD stuff on its result is plain stupid

ldjebran replied:
@rplevka this definitely not working with test discovered host as when provisioning the system want to reboot that host.

rplevka replied:
@ldjebran we don't do that in this test. Anyway, even if we did, UI handles this just fine and creates new host record from a fake discovered host just fine.

ldjebran replied:
suppose we have created a fake discovered host and we have this test code
```python
def test_positive_update_name(
        session, module_org, module_loc, module_host_group, discovered_host):
    """Update the discovered host name and provision it

    :id: 3770b007-5006-4815-ae03-fbd330aad304

    :Setup: Host should already be discovered

    :expectedresults: The hostname should be updated and host should be
        provisioned

    :CaseLevel: System
    """
    discovered_host_name = discovered_host['name']
    domain_name = module_host_group.domain.read().name
    new_name = gen_string('alpha').lower()
    new_host_name = '{0}.{1}'.format(new_name, domain_name)
    with session:
        session.organization.select(org_name=module_org.name)
        session.location.select(loc_name=module_loc.name)
        discovered_host_values = session.discoveredhosts.wait_for_entity(
            discovered_host_name)
        assert discovered_host_values['Name'] == discovered_host_name
        session.discoveredhosts.provision(
            discovered_host_name,
            module_host_group.name,
            module_org.name,
            module_loc.name,
            quick=False,
            host_values={'host.name': new_name}
        )
        assert (session.host.search(new_host_name)[0]['Name']
                == new_host_name)
        values = session.host.get_details(new_host_name)
        assert values['properties']['properties_table']['Status'] == 'OK'
        assert not session.discoveredhosts.search(
            'name = {0}'.format(discovered_host_name))
```
The test will fail in provisioning when trying to reboot the host here is a screen shot
![screenshot from 2018-10-23 13-30-22](https://user-images.githubusercontent.com/131553/47369456-64688f00-d6ec-11e8-8cd9-1579d9089c13.png)

rplevka commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,496 @@
+@run_only_on('sat')
+@skip_if_not_set('compute_resources', 'vlan_networking')
+@tier3
+def test_positive_delete(session, module_org, module_loc, provisioning_env):
```
similaral concern as the above one - using the mocked disco host, this will boil down to a simple CRUD which i would grade tier2 at max (if not tier1).

rplevka commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,496 @@
+@run_only_on('sat')
+@skip_if_not_set('compute_resources', 'vlan_networking')
+@tier3
+def test_positive_update_default_location(
```
\+ this will as well become t2 if fake host is used.

rplevka replied:
@ldjebran - i would simply generate a new uuid so it gets properly synced

ldjebran replied:
@rplevka after some experiments to join this 2 test come to that this will very complicate the tests that became unreadable , and hard to maintain , prefer to let them as is.

ldjebran replied:
now even if I want to do it I can't because of the bug, the test as is now , was tested to work.

rplevka commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,496 @@
+
+    :id: 3770b007-5006-4815-ae03-fbd330aad304
+
+    :Setup: Host should already be discovered
```
not a blocker - in order to honor what docstring says, we should have the discovered host already prepared by some fixture (setup).
this applies to all ocurrences of this

ldjebran commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,496 @@
+        return bool(result.return_code)
+
+
+def test_positive_read_details_from_discovered_host(
```
This test will not be run on automation it's here only to test that we are able to read the details page.

rplevka replied:
oh, right - the "test_" prefix confused me

rplevka commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,496 @@
+    default_discovery_auto = str(discovery_auto.value)
+    discovery_auto.value = 'True'
+    discovery_auto.update(['value'])
+    # we need the default RHEL os version
```
i don't think so - we don't use this at all. We do not do a proper provisioning anywhere in the below tests. So we can make some custom os, media and everything as we do it for `make_host()`

ldjebran replied:
@rplevka in the factory provisioning setup used here rely on creating a repo and choosing an os.

ldjebran replied:
@rplevka and if the os is wrongly chosen the factory provisioning setup will fail.

rplevka replied:
@ldjebran - my point was - you don't need to configure any provisioning at all, just don't call the factory.

ldjebran replied:
@rplevka I'am a bit confused here, as the goal of this module is to discover real hosts.

rplevka replied:
@ldjebran sorry for my last comment. Why don't you simply create some custom dummy repo + medium and use it?

you saide the following:
>  the factory provisioning setup used here rely on creating a repo and choosing an os.

but i'm confused - what factory are you talking about? can you point me to it?
i'm trying to say that syncing the real RH repo is a waste since we don't do any real provisioning anywhere => there's no real VM that would really need the packages.
All we test is that foreman creates the Host entry.

ldjebran replied:
As understood, the cli discovery tests are using medium and api tests are using repo + content view  that enable to use kickstart_repository (this scenario is not working with a dummy repo)
I'am receiving an error 
```
"error": {"id":null,"errors":{"name":[],"base":["The selected kickstart repository is not part of the assigned content view, lifecycle environment,\n                  content source, operating system, and architecture"]},"full_messages":["The selected kickstart repository is not part of the assigned content view, lifecycle environment,\n                  content source, operating system, and architecture"]}
```
CLI factory : https://github.com/SatelliteQE/robottelo/blob/master/robottelo/cli/factory.py#L3312
API factory: https://github.com/SatelliteQE/robottelo/blob/master/robottelo/api/utils.py#L298

Perhaps this was intentional by design to test the two scenarios.

rplevka replied:
@ldjebran,
ok Djebran, thanks for the explanation and sorry for holding this for long. i guess, let's proceed and merge the commit.
If I found ways of improving the tests, I'll submit a PR.

san7ket commented:
Ton of thanks on fixing this, unblocks my RHV PR Were you able to identify, why, on every cv.publish() and cv.promote() helper run we get a new duplicate repo added. I wanted to use the configure_provisining helper and got stuck on the similar problem

ldjebran commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,496 @@
+    discovery_auto = entities.Setting().search(
+        query={'search': 'name="discovery_auto"'})[0]
+    default_discovery_auto = str(discovery_auto.value)
+    discovery_auto.value = 'True'
```
any way if we remove this, we still have to change the discovery organization and location, so we have no escape from @run_in_one_thread.

ldjebran commented:
```
robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/ui_airgun/test_discoveredhost.py 
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: file
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.2.1, mock-1.6.3, forked-0.2, cov-2.5.1
collecting 8 items                                                                                           2018-09-18 14:58:02 - conftest - DEBUG - BZ deselect is disabled in settings

collected 8 items                                                                                            

tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_read_details_from_discovered_host PASSED [ 12%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_pxe_based_discovery PASSED               [ 25%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_pxe_less_with_dhcp_unattended PASSED     [ 37%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_provision_using_quick_host_button PASSED [ 50%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_name PASSED                       [ 62%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_manual_provision_host_with_rule PASSED   [ 75%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_delete PASSED                            [ 87%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_reboot PASSED                            [100%]

======================================== 8 passed in 1333.56 seconds =========================================
```
assign location and organization has been affected by different bugs in last snap

ldjebran commented:
remove do not merge as dependency merged

ldjebran commented:
removed test for read discovered host as without doc string and id

rplevka commented:
done, comments added

ldjebran commented:
Moved some tests to tier2 using fake discovered hosts
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m "tier2" tests/foreman/ui_airgun/test_discoveredhost.py
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: file
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.0, mock-1.10.0, forked-0.2, cov-2.5.1
collecting 8 items                                                                                           2018-10-23 18:42:16 - conftest - DEBUG - BZ deselect is disabled in settings

collected 8 items / 4 deselected                                                                             

tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_provision_using_quick_host_button PASSED [ 25%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_manual_provision_host_with_rule PASSED   [ 50%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_delete PASSED                            [ 75%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_default_taxonomies SKIPPED        [100%]

============================ 3 passed, 1 skipped, 4 deselected in 276.71 seconds =============================

```

ldjebran commented:
Moved some tests to tier2 using fake discovered hosts and joined the taxonomies tests
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m "tier2" tests/foreman/ui_airgun/test_discoveredhost.py
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: file
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.0, mock-1.10.0, forked-0.2, cov-2.5.1
collecting 8 items                                                                                           2018-10-23 19:08:15 - conftest - DEBUG - BZ deselect is disabled in settings

collected 8 items / 3 deselected                                                                             

tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_provision_using_quick_host_button PASSED [ 20%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_name PASSED                       [ 40%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_manual_provision_host_with_rule PASSED   [ 60%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_delete PASSED                            [ 80%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_default_taxonomies SKIPPED        [100%]

============================ 4 passed, 1 skipped, 3 deselected in 421.97 seconds =============================
```

ldjebran commented:
@rplevka @oshtaier changes done.

Waiting this PR to be merged to continue porting tests for discovered hosts.
@oshtaier Think we can mark that issue https://github.com/SatelliteQE/robottelo/issues/6372 as blocked by this PR.

oshtaier commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,414 @@
+    with LibvirtGuest() as pxe_host:
+        host_name = pxe_host.guest_name
+        with session:
+            session.organization.select(org_name=module_org.name)
```
it is redundant across all test cases

oshtaier commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,414 @@
+        host_name = pxe_host.guest_name
+        with session:
+            session.organization.select(org_name=module_org.name)
+            session.location.select(loc_name=module_loc.name)
```
maybe we can implement something similar for `module_loc` as we have for `module_org` and remove all the statements

oshtaier commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,414 @@
+            assert _is_host_reachable(host_ip)
+            session.discoveredhosts.apply_action('Reboot', host_name)
+            # Ensure that the host is not reachable
+            assert not _is_host_reachable(host_ip, expect_reachable=False)
```
emm. Shouldn't we wait till host become reachable to actually verify 'Host should be successfully rebooted.'?

ldjebran replied:
Think it's sufficient to check that the command passed and the host became unreachable (as the original version do), but if you want we can wait for that host to became  reachable again.

ldjebran replied:
But unfortunately after some experiment seems the host is completely deleted from the infra after reboot perhaps this the infra configuration to keep only the running vms:
so if our VM name macb68e5d3620db.XXXX.com

before reboot
```
virsh # list
 Id    Name                           State
----------------------------------------------------
 13869 macb68e5d3620db.XXXX.com running
```

just after reboot command from sat

```
virsh # list
 Id    Name                           State
----------------------------------------------------
 13869 macb68e5d3620db.XXXX.com shut off
```

after this step the VM completely disappear from the VMs list

any way we are not testing the libvirt and abaility to do reboot , but that the command was transmitted to infra.

after reboot the VM also fail to be deleted as does not exist any more
from robottelo log
```
virsh destroy macb68e5d3620db.XXXX.com
rror: Failed to destroy domain macb68e5d3620db.satellite.lab.eng.rdu2.redhat.com
error: Requested operation is not valid: domain is not running
```

ldjebran replied:
Think no as per investigations we made  with @san7ket it seems that the VM is shutdown it's not deleted it simply not listed as running and cannot reboot as it has not been provisioned, the boot loader must be configured at provisioning stage and as no provisioning took place, there was no boot media to boot from.

ldjebran replied:
so the VM after reboot was intended to boot from it's own media, but that media was not provisioned.

oshtaier commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,414 @@
+            location=[module_loc],
+            organization=[module_org],
+        ).create()
+        session.discoveredhosts.apply_action(
```
why title is `manual_provision`, but we do `Auto Provision? It is just not logical. I feel you are doing right things, so maybe we just need to rephrase title or?

oshtaier commented on tests/foreman/ui_airgun/test_discoveredhost.py:
```diff
@@ -0,0 +1,414 @@
+        host_name = ('{0}.{1}'.format(discovered_host_name, domain.name))
+        assert (session.host.search(host_name)[0]['Name'] == host_name)
+        values = session.host.get_details(host_name)
+        assert values['properties']['properties_table']['Status'] == 'OK'
```
is it possible to verify in any way that discovery rule was actually used in the flow?

ldjebran commented:
@oshtaier addressed comments
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m 'tier2' tests/foreman/ui_airgun/test_discoveredhost.py 
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: file
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.0, mock-1.10.0, forked-0.2, cov-2.5.1
collecting 8 items                                                                                           2018-11-12 18:25:23 - conftest - DEBUG - BZ deselect is disabled in settings

collected 8 items / 3 deselected                                                                             

tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_provision_using_quick_host_button PASSED [ 20%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_name PASSED                       [ 40%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_auto_provision_host_with_rule PASSED     [ 60%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_delete PASSED                            [ 80%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_default_taxonomies SKIPPED        [100%]

============================ 4 passed, 1 skipped, 3 deselected in 399.19 seconds =============================
```
